stl/inc/xstring

```diff
@@ -56,11 +56,15 @@ struct _Char_traits { // properties of a string or stream element
 
         return _Count;
     }
-
-    static _Elem* copy(_Out_writes_all_(_Count) _Elem* const _First1, _In_reads_(_Count) const _Elem* _First2,
-        size_t _Count) noexcept /* strengthened */ {
+    static _CONSTEXPR20_MEMOP _Elem* copy(_Out_writes_all_(_Count) _Elem* const _First1,
+        _In_reads_(_Count) const _Elem* const _First2, const size_t _Count) noexcept /* strengthened */ {

```

Comment:
Note to self: I will investigate avoiding `_CONSTEXPR20_MEMOP` (and fixing the remaining skipped tests?) by using `is_constant_evaluated` as a fallback.

Reply:
Originally we did that. However, the problem is that to not crash users with overlapping ranges we would need to compare two iterators that might be from different arrays. 

Maybe comparing the unwrapped iterators would be the better choice.

Also to note. The comparison is only necessary to support technically UB usage of copy as overlapping ranges are explicitely forbidden

Reply:
Thanks, I see the discussion now (still getting used to keeping up with GitHub's unstructured comments outside of code review lines).

I'm going to try a hilariously simple hack - we should be able to compare pointers for *equality* even if we can't compare for *less-than*, so when `is_constant_evaluated`, just blast through one of the ranges in linear time and check for overlap at every location! (Linear time is constant time, thanks `constexpr`.) If that doesn't work, I will be :crying_cat_face:

Reply:
That would even be efficient because you can essentially do it while copying.

Ship it!

Reply:
I discovered that LWG-3085 "`char_traits::copy` precondition too weak" was filed in March 2018. The precondition as written in the WP (permitting overlap in the opposite direction as `std::copy`) makes no sense; I believe that the intention was always to forbid any overlap and to permit directly calling `memcpy`. I prefer the Proposed Resolution and have told the LWG so.

So, I am going to try the following:
* call the builtin when available,
* otherwise use `is_constant_evaluated` to call `memcpy/memmove` at runtime,
* otherwise at compiletime, *don't* check for overlap in `copy`, do check for any overlap in `move`.

Reply:
Testing this now. Note that we only have to check for overlap in one direction (the direction forbidden by `std::copy`), but we have to perform the check as a separate initial pass - we can't start copying and then discover overlap, because it'll be too late to salvage the wreckage.